"." columns

I read your post ["Efficient Tabular Storage"] (http://matthewrocklin.com/blog/work/2015/08/28/Storage/)  and decided to to give Castra a whirl.  As my luck would have it, I got an error on the very first data set I tried.

The problem is that the dataframe has a column that is labeled "." (period, without quotes) and, of course, on *nix this is an invalid name for regular files.  I don't know if/how you would want to fix this.  I've written a test case nontheless.

    def test_column_with_period():                                                                      
        df = pd.DataFrame({'x': [10, 20],                                                               
                           '.': [10., 20.]},                                                            
                          columns=['x', '.'],                                                           
                          index=[10, 20])                                                               
                                                                                                        
        with Castra(template=df) as c:                                                                  
            c.extend(df) 

Which yields:

    >       with open(filename, 'wb') as fp:
    E       IOError: [Errno 21] Is a directory: '/tmp/castra-uUz3Gr/10--20/.'

--- a/castra/core.py
+++ b/castra/core.py
@@ -51,7 +51,7 @@ def escape(text):
     """
     result = ''.join(c for c in str(text) if c in valid_chars)
     if not result:
-        result = md5(str(text)).hexdigest()
+        result = md5(str(text).encode()).hexdigest()
     return result
 
 
@@ -234,7 +234,7 @@ class Castra(object):
             self.extend(buf)
 
     def dirname(self, *args):
-        return os.path.join(self.path, *map(escape, args))
+        return os.path.join(self.path, *list(map(escape, args)))
 
     def load_partition(self, name, columns, categorize=True):
         if isinstance(columns, Iterator):

--- a/castra/core.py
+++ b/castra/core.py
@@ -36,8 +36,13 @@ def blosc_args(dt):
     return None
 
 
+replacements = {'.': '-dot', '/': '-slash-', '\\': '-backslash-', ' ': ''}
+
 def escape(text):
-    return str(text)
+    text = str(text)
+    for k, v in replacements.items():
+        text = text.replace(k, v)
+    return text
 
 
 def mkdir(path):
@@ -219,7 +224,7 @@ class Castra(object):
             self.extend(buf)
 
     def dirname(self, *args):
-        return os.path.join(self.path, *args)
+        return os.path.join(self.path, *map(escape, args))
 
     def load_partition(self, name, columns, categorize=True):
         if isinstance(columns, Iterator):

--- a/castra/core.py
+++ b/castra/core.py
@@ -36,13 +36,23 @@ def blosc_args(dt):
     return None
 
 
-replacements = {'.': '-dot', '/': '-slash-', '\\': '-backslash-', ' ': ''}
+# http://stackoverflow.com/questions/295135/turn-a-string-into-a-valid-filename-in-python
+import string
+valid_chars = "-_%s%s" % (string.ascii_letters, string.digits)
 
 def escape(text):
-    text = str(text)
-    for k, v in replacements.items():
-        text = text.replace(k, v)
-    return text
+    """
+
+    >>> escape("Hello!")  # Remove punctuation from names
+    'Hello'
+
+    >>> escape("/!.")  # completely invalid names produce hash string
+    'cb6698330c63e87fc35933a0474238b0'
+    """
+    result = ''.join(c for c in str(text) if c in valid_chars)
+    if not result:
+        result = md5(str(text)).hexdigest()
+    return result
 
 
 def mkdir(path):

--- a/castra/tests/test_core.py
+++ b/castra/tests/test_core.py
@@ -444,3 +444,12 @@ def test_extend_sequence_overlap():
         c.extend_sequence(seq)
         tm.assert_frame_equal(c[:], df)
         assert (c.partitions.index == [9, 16]).all()
+
+def test_column_with_period():
+    df = pd.DataFrame({'x': [10, 20],
+                       '.': [10., 20.]},
+                       columns=['x', '.'],
+                       index=[10, 20])
+
+    with Castra(template=df) as c:
+        c.extend(df)

